forc-util/src/lib.rs

```diff
@@ -407,6 +410,41 @@ fn construct_window<'a>(
     &input[calculated_start_ix..calculated_end_ix]
 }
 
+const LOG_FILTER: &str = "RUST_LOG";
+const HUMAN_LOGGING: &str = "HUMAN_LOGGING";
+
+pub fn set_subscriber() {
+    let filter = match env::var_os(LOG_FILTER) {
+        Some(_) => EnvFilter::try_from_default_env().expect("Invalid `RUST_LOG` provided"),
+        None => EnvFilter::new("info"),
+    };
+
+    let human_logging = env::var_os(HUMAN_LOGGING)
```

Comment:
I think it might be worth leaving `HUMAN_LOGGING` out of this PR in favour of introducing it in a follow-up PR. I think there's still some open Qs around how we want this to behave. Some Qs that come to mind:

- Rather than `HUMAN_LOGGING`, should this be `JSON_LOGGING`? This might be more flexible if we need to support more formats in the future?
- If we do want an option for machine-readable JSON formatting (which I think we do), this should provide more structured error/warning messages too, rather than only representing the outer tracing events as JSON with a large string blob under `message` in `fields`. E.g. this is the output I get using `HUMAN_LOGGING=false` when building one of the tests that emits some warnings:

  ```
    Running `target/debug/forc build --path ./test/src/e2e_vm_tests/test_programs/should_pass/language/array_basics/`
  {"timestamp":"2022-05-19T04:05:51.929438Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:5:12\n\u001b[1;38;5;12m  |\u001b[0m\n\u001b[1;38;5;12m3 |\u001b[0m \n\u001b[1;38;5;12m4 |\u001b[0m     /// The smallest value that can be represented by this integer type.\n\u001b[1;38;5;12m5 |\u001b[0m     pub fn MIN() -> u64 {\n\u001b[1;38;5;12m  |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MIN\" is not idiomatic. Function names should be snake_case, like \"min\".\u001b[0m\n\u001b[1;38;5;12m6 |\u001b[0m         0\n\u001b[1;38;5;12m7 |\u001b[0m     }\n\u001b[1;38;5;12m  |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.929654Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:11:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m 9 |\u001b[0m \n\u001b[1;38;5;12m10 |\u001b[0m     /// 2<sup>64</sup> - 1.\n\u001b[1;38;5;12m11 |\u001b[0m     pub fn MAX() -> u64 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MAX\" is not idiomatic. Function names should be snake_case, like \"max\".\u001b[0m\n\u001b[1;38;5;12m12 |\u001b[0m         18446744073709551615\n\u001b[1;38;5;12m13 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.929843Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:16:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m14 |\u001b[0m \n\u001b[1;38;5;12m15 |\u001b[0m     /// The size of this integer type in bits.\n\u001b[1;38;5;12m16 |\u001b[0m     pub fn BITS() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ----\u001b[0m \u001b[1;38;5;11mFunction name \"BITS\" is not idiomatic. Function names should be snake_case, like \"bits\".\u001b[0m\n\u001b[1;38;5;12m17 |\u001b[0m         64\n\u001b[1;38;5;12m18 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.930047Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:23:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m21 |\u001b[0m \n\u001b[1;38;5;12m22 |\u001b[0m     /// The smallest value that can be represented by this integer type.\n\u001b[1;38;5;12m23 |\u001b[0m     pub fn MIN() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MIN\" is not idiomatic. Function names should be snake_case, like \"min\".\u001b[0m\n\u001b[1;38;5;12m24 |\u001b[0m         0\n\u001b[1;38;5;12m25 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.930256Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:29:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m27 |\u001b[0m \n\u001b[1;38;5;12m28 |\u001b[0m     /// 2<sup>32</sup> - 1.\n\u001b[1;38;5;12m29 |\u001b[0m     pub fn MAX() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MAX\" is not idiomatic. Function names should be snake_case, like \"max\".\u001b[0m\n\u001b[1;38;5;12m30 |\u001b[0m         4294967295\n\u001b[1;38;5;12m31 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.930478Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:34:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m32 |\u001b[0m \n\u001b[1;38;5;12m33 |\u001b[0m     /// The size of this integer type in bits.\n\u001b[1;38;5;12m34 |\u001b[0m     pub fn BITS() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ----\u001b[0m \u001b[1;38;5;11mFunction name \"BITS\" is not idiomatic. Function names should be snake_case, like \"bits\".\u001b[0m\n\u001b[1;38;5;12m35 |\u001b[0m         32\n\u001b[1;38;5;12m36 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.930705Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:41:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m39 |\u001b[0m \n\u001b[1;38;5;12m40 |\u001b[0m     /// The smallest value that can be represented by this integer type.\n\u001b[1;38;5;12m41 |\u001b[0m     pub fn MIN() -> u16 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MIN\" is not idiomatic. Function names should be snake_case, like \"min\".\u001b[0m\n\u001b[1;38;5;12m42 |\u001b[0m         0\n\u001b[1;38;5;12m43 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.930946Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:47:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m45 |\u001b[0m \n\u001b[1;38;5;12m46 |\u001b[0m     /// 2<sup>16</sup> - 1.\n\u001b[1;38;5;12m47 |\u001b[0m     pub fn MAX() -> u16 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MAX\" is not idiomatic. Function names should be snake_case, like \"max\".\u001b[0m\n\u001b[1;38;5;12m48 |\u001b[0m         65535\n\u001b[1;38;5;12m49 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.931192Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:52:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m50 |\u001b[0m \n\u001b[1;38;5;12m51 |\u001b[0m     /// The size of this integer type in bits.\n\u001b[1;38;5;12m52 |\u001b[0m     pub fn BITS() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ----\u001b[0m \u001b[1;38;5;11mFunction name \"BITS\" is not idiomatic. Function names should be snake_case, like \"bits\".\u001b[0m\n\u001b[1;38;5;12m53 |\u001b[0m         16\n\u001b[1;38;5;12m54 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.931450Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:59:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m57 |\u001b[0m \n\u001b[1;38;5;12m58 |\u001b[0m     /// The smallest value that can be represented by this integer type.\n\u001b[1;38;5;12m59 |\u001b[0m     pub fn MIN() -> u8 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MIN\" is not idiomatic. Function names should be snake_case, like \"min\".\u001b[0m\n\u001b[1;38;5;12m60 |\u001b[0m         0\n\u001b[1;38;5;12m61 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.931718Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:65:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m63 |\u001b[0m \n\u001b[1;38;5;12m64 |\u001b[0m     /// 2<sup>8</sup> - 1.\n\u001b[1;38;5;12m65 |\u001b[0m     pub fn MAX() -> u8 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ---\u001b[0m \u001b[1;38;5;11mFunction name \"MAX\" is not idiomatic. Function names should be snake_case, like \"max\".\u001b[0m\n\u001b[1;38;5;12m66 |\u001b[0m         255\n\u001b[1;38;5;12m67 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
  {"timestamp":"2022-05-19T04:05:51.932002Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m sway-lib-core/src/num.sw:70:12\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m68 |\u001b[0m \n\u001b[1;38;5;12m69 |\u001b[0m     /// The size of this integer type in bits.\n\u001b[1;38;5;12m70 |\u001b[0m     pub fn BITS() -> u32 {\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m            ----\u001b[0m \u001b[1;38;5;11mFunction name \"BITS\" is not idiomatic. Function names should be snake_case, like \"bits\".\u001b[0m\n\u001b[1;38;5;12m71 |\u001b[0m         8\n\u001b[1;38;5;12m72 |\u001b[0m     }\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
    Compiled library "core" with 12 warnings.
  {"timestamp":"2022-05-19T04:05:51.963508Z","level":"ERROR","fields":{"message":"\u001b[1;38;5;11mwarning\u001b[0m\n  \u001b[1;38;5;12m-->\u001b[0m array_basics/src/main.sw:10:24\n\u001b[1;38;5;12m   |\u001b[0m\n\u001b[1;38;5;12m 8 |\u001b[0m \n\u001b[1;38;5;12m 9 |\u001b[0m     let a: [bool; 5] = [true, true, true, false, true];\n\u001b[1;38;5;12m10 |\u001b[0m     let b: [u32; 10] = [3; 10];\n\u001b[1;38;5;12m   |\u001b[0m\u001b[1;38;5;11m                        -------\u001b[0m \u001b[1;38;5;11mThis cast, from integer type of width sixty four to integer type of width thirty two, will lose precision.\u001b[0m\n\u001b[1;38;5;12m11 |\u001b[0m     let c = [0x01, 0x02, 0x03];\n\u001b[1;38;5;12m12 |\u001b[0m     let d = [0; 10];\n\u001b[1;38;5;12m   |\u001b[0m\n____\n"},"target":"forc_util","line_number":358}
    Compiled script "array_basics" with 1 warning.
  {"timestamp":"2022-05-19T04:05:51.964110Z","level":"INFO","fields":{"message":"  Bytecode size is 1380 bytes."},"target":"forc::ops::forc_build","line_number":124}
  ```
  ^ The `Compiled library "core" with 12 warnings.` and `Compiled script "array_basics" with 1 warning` lines look like we might also be missing some cases where we should use `tracing` but aren't yet?

Reply:
I am removing HUMAN_LOGGING for this PR and will look into if we miss somewhere for `tracing`.